### PR TITLE
Front service with ELB with scheduled instance rotation

### PR DESCRIPTION
**Pull request by spohnan**
**Use case:** Development account where many developers access via the bastion. 

**Additional Features:**
* Single named point of entry _bastion.projectX.mycompany.com_
  * Server certs are synced using a private S3 bucket
* Auto rotation of bastion instances using ASG ScheduledActions
  * Example is excessively fast just to demonstrate but Recurrence would become a parameter
  * Additional level of difficulty to maintain a entry point if bastion becomes compromised

**Cons**
* S3 bucket read/write in BastionHostRole increases exposure of server certs to users with local accounts.

I realize this probably isn't a common enough use case to merge this into the quickstart repo but if you have any feedback at all about the implementation I'd be really interested to hear your thoughts.

### CODE REVIEW

**Comment by santiagocardenas** [comment]
Hi @spohnan This looks like some nice additions.

I'm open to merge these if we retain existing functionality. That is, making the ELB-based architecture an option (thus not using EIPs). EIPs are nice because they are predictable for corporate firewalls to manage rules against.

Also, the certificate piece should support both BYO cert as well as AWS Certificate Manager (free certs!).

Thanks,
Santiago

**Comment by spohnan** [comment]
Thanks for taking a look @santiagocardenas  I've made updates to incorporate your suggestions:

* Added ELB option with the EIP deployment remaining the default
* Instance auto-rotation behind ELB is off by default but available by configuring two cron expressions (scale-up and scale-down which should overlap slightly)
* Both options now take advantage of the new [aws ec2 replace-iam-instance-profile-association](https://aws.amazon.com/blogs/security/new-attach-an-aws-iam-role-to-an-existing-amazon-ec2-instance-by-using-the-aws-cli/) call to drop from a more privileged initialization role to an operating role that is just allowed to push logs. This addresses my initial concern about exposing the bastion server SSH host certs to local users that could figure out the location of the S3 bucket.

Notes:

* ELB timeout has been bumped up to 75 seconds but users would need to ensure they have a ServerAliveInterval configured to prevent their session from constantly timing out every minute
* The docs would have to be updated a bit to explain the additional option, when each option would be the best choice and show the (optional) ELB/S3 resources in the architecture diagram
* I'm willing to help with content update suggestions and/or making additional updates to the code if you see issues

Question:

* I'm familiar with ACM and think it's a wonderful service but we're just running SSH on the bastion not SSL/TLS so I didn't really understand your last suggestion

**Comment by santiagocardenas** [comment]
@spohnan, 
So yes, I was thinking of something else when I suggested using ACM. I'll have some colleagues take a look at these changes as well. A few things that I see:
- The output of the BastionSecurityGroupID has been removed. This is needed with that exact name as it's consumed by other Quick Starts
- Port TCP/443 ingress the ELB security group, but you're just using SSH TCP/22. This ingress rule should be removed from the security group.

We'll also totally take you up on helping out with updating the docs and greatly appreciate your contributions!

In the meantime, please confirm that these changes and any future documentationare under the Apache 2.0 license terms.

Thanks,
Santiago

File: templates/linux-bastion.template [hunk 1]
```diff
@@ -849,12 +1282,6 @@
                 "Ref": "BastionMainLogGroup"
             },
             "Description": "CloudWatch Logs GroupName. Your SSH logs will be stored here."
-        },
-        "BastionSecurityGroupID": {
```
Comment (santiagocardenas) [inline templates/linux-bastion.template]: This is needed. I would suggest to combine the BastionSecurityGroup and BastionELBSecurityGroup into a single resource BastionSecurityGroup and use condition functions to determine the ingress rules properties.

File: templates/linux-bastion.template [hunk 2]
```diff
@@ -797,11 +1190,49 @@
...
+                        "ToPort": "22",
+                        "CidrIp": {
+                            "Ref": "RemoteAccessCIDR"
+                        }
                     },
                     {
-                        "IpProtocol": "icmp",
-                        "FromPort": "-1",
-                        "ToPort": "-1",
+                        "IpProtocol": "tcp",
```
Comment (santiagocardenas) [inline templates/linux-bastion.template]: Why is 443 allowed?

**Comment by spohnan** [comment]
Changes made after first review:
----
* BastionAutoScalingGroup and BastionSecurityGroupID Outputs restored
* Port 443 ingress removed from BastionLoadBalancer and ELBSecurityGroup
* ICMP restored to BastionSecurityGroup
* Merged BastionSecurityGroup and BastionELBSecurityGroup
* Merged BastionAutoScalingGroup and BastionELBAutoScalingGroup

[Documentation update suggestions](https://github.com/spohnan/quickstart-linux-bastion-docs/compare/dev)
----
* Not a perfect copy of the doc source. I just exported from PDF and cleaned up the resulting markup a bit
* Meant to highlight diffs

License
----
* The patch and suggested documentation updates are both licensed under the Apache License, Version 2.0 (the "License")

**Pull request closed without merging**